### PR TITLE
[Debugger] Debugger failure when field content is assigned by an unsafe value

### LINKED ISSUE #12374: Unable to inspect ReadOnlySpan<char> in debugger
## Steps to Reproduce

1. Have an application with ReadOnlySpan being used
```
var span = "asdf".AsSpan();
```
2. Given a StructMirror invoke of `ReadOnlySpan<char>.ToString()`, it'll fail with an `INVALID_ARGUMENT` error code.
3. Maybe ref structs need to be specialized in the debugger when passed?

### Current Behavior

We should get the string.

### Expected Behavior

INVALID_ARGUMENT error code

## On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**:

```
Mono JIT compiler version 5.18.0.223 (2018-08/3f799cc3c78 Wed Dec 19 22:12:52 EST 2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug
	Interpreter:   yes
	LLVM:          yes(600)
	Suspend:       preemptive
	GC:            sgen (concurrent by default)
```

## Stacktrace

```
Please paste the stack trace here if available.
```

### PR DESCRIPTION
**Pull request by thaystg**
Debugger failure when a method is called in watches and any field of the Class has the type of the field different of the value of the field because the assign was done in an unsafe way like this:

Now the datatypes check are only executed when it's not inside an invoke method.
_pinnable = Unsafe.As<Pinnable<T>>(array);
A unit test that reproduces the issue was created either.
Fixes #12374

### CODE REVIEW

File: mono/mini/debugger-agent.c [hunk 1]
```diff
@@ -5543,8 +5543,11 @@ decode_value_internal (MonoType *t, int type, MonoDomain *domain, guint8 *addr,
 
 				if (obj) {
 					if (!obj_is_of_type (obj, t)) {
-						DEBUG_PRINTF (1, "Expected type '%s', got '%s'\n", mono_type_full_name (t), m_class_get_name (obj->vtable->klass));
-						return ERR_INVALID_ARGUMENT;
+						DebuggerTlsData *tls = (DebuggerTlsData *)mono_native_tls_get_value (debugger_tls_id);
```
[the remark this answers is not in the document]
Comment (thaystg) [reply mono/mini/debugger-agent.c]: If the object doesn't implement the method or if the method has a throw it's not there the check. 
As I understood this function is only called when you are running a method before the method execution for each field of the class.
It's called from this line: debugger-agent.c:6086
And it's called too for each of the parameters passed to the method. I think that the parameters have to be checked(and not check is unsafe in this case), so I'm thinking about change the signature of the method to receive a parameter if we want to check or not the dataypes.

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
I guess this is ok. However, if we want to be able to debug all the weird stuff which can be created by unsafe code, we would eventually have to turn off a lot of checks.

**Comment by marek-safar** [comment]
@monojenkins backport 2018-12

**Comment by marek-safar** [comment]
@thaystg could you backport it manually to 2018-12

**Comment by thaystg** [comment]
Yes, of course! :)

**Pull request merged**